src/mpi/init/init_async.c

```diff
@@ -59,22 +74,163 @@ static void progress_fn(void *data)
     return;
 }
 
+#if defined(MPL_HAVE_MISC_GETNPROCS)
+MPL_STATIC_INLINE_PREFIX int MPIDI_parse_progress_thread_affinity(int *thread_affinity,
+                                                                  int async_threads_per_node)
+{
+    int th_idx, read_count = 0, mpi_errno = MPI_SUCCESS;
+    char *affinity_copy = NULL;
+    const char *affinity_to_parse = MPIR_CVAR_CH4_PROGRESS_THREAD_AFFINITY;
+    char *proc_id_str, *tmp;
+    size_t proc_count;
+
+    if (!affinity_to_parse || strlen(affinity_to_parse) == 0) {
+        printf("ERROR: unexpected affinity string\n");
+        mpi_errno = MPI_ERR_OTHER;
+        goto fn_fail;
+    }
+
+    /* create copy of original buffer because it will be modified in strsep */
+    affinity_copy = MPL_strdup(affinity_to_parse);
+    MPIR_Assert(affinity_copy);
+    tmp = affinity_copy;
+
+    /* apply auto affinity */
+    if (!strcmp(affinity_copy, "auto")) {
+        /* generate default affinity */
+        proc_count = MPL_get_nprocs();
+        for (th_idx = 0; th_idx < async_threads_per_node; th_idx++) {
+            if (th_idx < proc_count)
+                thread_affinity[th_idx] = proc_count - th_idx - 1;
+            else
+                thread_affinity[th_idx] = thread_affinity[th_idx % proc_count];
```

Comment:
You should do `proc_count - (th_idx % proc_count) - 1` for (ordering) consistency.

Reply:
Done